src/Features/Core/Portable/Extensions/IExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,73 @@
+    /// <param name="assemblyFilePath">The assembly to register and create message handlers from.</param>
+    /// <param name="cancellationToken">Cancellation token to cancel the async operation.</param>
+    /// <returns>The names of the registered handlers.</returns>
+    ValueTask<RegisterExtensionResponse> RegisterExtensionAsync(
+        Solution solution,
+        string assemblyFilePath,
+        CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Executes a non-document-specific extension message handler with the given message and solution.
+    /// </summary>
+    /// <param name="solution">The solution the message refers to.</param>
+    /// <param name="messageName">The name of the handler to execute. This is generally the full name of the type implementing the handler.</param>
+    /// <param name="jsonMessage">The json message to be passed to the handler.</param>
+    /// <param name="cancellationToken">Cancellation token to cancel the async operation.</param>
+    /// <returns>The json message returned by the handler.</returns>
+    ValueTask<string> HandleExtensionWorkspaceMessageAsync(
+        Solution solution,
+        string messageName,
+        string jsonMessage,
+        CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Executes a document-specific extension message handler with the given message and solution.
+    /// </summary>
+    /// <param name="documentId">The document the message refers to.</param>
+    /// <param name="messageName">The name of the handler to execute. This is generally the full name of the type implementing the handler.</param>
+    /// <param name="jsonMessage">The json message to be passed to the handler.</param>
+    /// <param name="cancellationToken">Cancellation token to cancel the async operation.</param>
+    /// <returns>The json message returned by the handler.</returns>
+    ValueTask<string> HandleExtensionDocumentMessageAsync(
+        Document documentId,
+        string messageName,
+        string jsonMessage,
+        CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Unregisters extension message handlers previously registered from <paramref name="assemblyFilePath"/>.
+    /// </summary>
+    /// <param name="assemblyFilePath">The assembly for which handlers should be unregistered.</param>
+    /// <param name="cancellationToken">Cancellation token to cancel the async operation.</param>
+    /// <returns>A task representing the async operation.</returns>
+    ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Unregisters all extension message handlers.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token to cancel the async operation.</param>
+    /// <returns>A task representing the async operation.</returns>
+    ValueTask ResetAsync(
+        CancellationToken cancellationToken);
```

Comment:
open questions from above:

1. why even have this?
2. why is it async?
3. why is it cancellable?

Reply:
ok.  This needs this because it *should* be calling into teh IRemoteXXXService to do the work (which would then call back into this on the OOP side).  So this should be async/cancellable.  BUt we still want to make it so that the impl of this calls to remote.  not the handler itself.

Reply:
I changed it to Reset()